docs/architecture/adr-030-msg-auth-module.md

```diff
@@ -0,0 +1,222 @@
+# ADR 030: Msg Authorization Module
+
+## Changelog
+
+- 2019-11-06: Initial Draft
+- 2020-10-12: Updated Draft
+
+## Status
+
+Accepted
+
+## Abstract
+
+This ADR defines the `msg_authorization` module which allows accounts to grant authorizations to perform actions
+on behalf of that account to other accounts.
+
+## Context
+
+The concrete use cases which motivated this module include:
+- the desire to delegate the ability to vote on proposals to other accounts besides the account which one has
+delegated stake
+- "sub-keys" functionality which is a term used to describe the functionality provided by this  module together with
+the `fee_grant` module from [ADR 029](./adr-029-fee-grant-module.md) and the [group module](https://github.com/regen-network/cosmos-modules/tree/master/incubator/group).
+
+The "sub-keys" functionality roughly refers to the ability for one account to grant some subset of its capabilities to
+other accounts with possibly less robust, but easier to use security measures. For instance, a master account representing
+an organization could grant the ability to spend small amounts of the organization's funds to individual employee accounts.
+Or an individual (or group) with a multisig wallet could grant the ability to vote on proposals to any one of the member
+keys.
+
+The sub-keys functionality was originally proposed in https://github.com/cosmos/cosmos-sdk/issues/4480. The current
+implementation is based on work done by the [Gaian's team at Hackatom Berlin 2019](https://github.com/cosmos-gaians/cosmos-sdk/tree/hackatom/x/delegation).
+
+## Decision
+
+We will create a module named `msg_authorization` which provides support for
+granting arbitrary capabilities from one account (the granter) to another account (the grantee). Authorizations
+must be granted for a particular `Msg` service methods one by one using an implementation
+of `Authorization`.
+
+### Types
+
+Authorizations determine exactly what capabilities are granted. They are extensible
+and can be defined for any `Msg` service method even outside of the module where
+the `Msg` method is defined. `Authorization`s use the new `ServiceMsg` type from
+ADR 031.
+
+#### Authorization
+
+```go
+type Authorization interface {
+	// MethodName returns the fully-qualified Msg service method name as described in ADR 031.
+	MethodName() string
+
+	// Accept determines whether this grant permits the provided sdk.ServiceMsg to be performed, and if
+	// so provides an upgraded authorization instance.
+	Accept(msg sdk.ServiceMsg, block abci.Header) (allow bool, updated Authorization, delete bool)
```

Comment:
Why do we need the last parameter: `delete`? It's not documented anywhere?

Reply:
It's for when an authorization has been "exhausted". For instance the fully allowed spend limit has been spent. I'll update the docs 👍

Reply:
Do you think it's useful to have it? Probably we should also be able to prune old authorizations to save the space (this can grow). People probably should stake some tokens to create an authorization to avoid growing the ledger. This will be returned when someone deletes the authorization. If someone will like to do an audit, then he can still check the ledger.

Reply:
Well this `delete` flag automatically prunes

Reply:
> People probably should stake some tokens to create an authorization to avoid growing the ledger. 

Let's add a gas cost here.

Also `delete` won't cause pruning if the authorization isn't successful so yes there should be a separate pruning process. Any suggestions on how we do that @robert-zaremba ?

Reply:
Created this issue to follow up on this: https://github.com/cosmos/cosmos-sdk/issues/8311